rfd/0151-database-permission-management.md

```diff
@@ -0,0 +1,355 @@
+---
+authors: Krzysztof Skrzętnicki (krzysztof.skrzetnicki@goteleport.com)
+state: draft
+---
+
+# RFD 151 - Database Permission Management
+
+## Required Approvals
+
+- Engineering: @r0mant && @smallinsky
+- Product: @xinding33 || @klizhentas
+- Security: @reedloden || @jentfoo
+
+## What
+
+Extends [automated db user provisioning](0113-automatic-database-users.md) with
+permission management capabilities.
+
+## Why
+
+Managing database-level permissions is a natural extension of current Teleport
+RBAC capabilities. The described model should also integrate seamlessly with
+TAG, providing a future-proof solution.
+
+Database administrators will be able to use Teleport for both user and
+permission management.
+
+## Details
+
+### UX
+
+A set of global "database object import rules" resources are defined for the
+Teleport cluster. At certain points in time, the database schema is read and
+passed through the import rules. The individual import rules may apply custom
+labels. An database object resource will only be created if there is at least
+one import rule that matched it.
+
+Only object attributes (standard ones like `protocol` or custom from the
+`attributes` field), which are sourced from the object spec, are subject to
+matching against the import rules. Any labels present on an object, either from
+an import rule or another source, are not subject to be matched by the import
+rule. The permissions matching is different in this regard.
+
+Example: the `sales-prod` import rule, which imports all tables from `sales`
+Postgres database in prod:
+
+```yaml
+kind: db_object_import_rule
+metadata:
+  name: sales-prod
+spec:
+  db_labels:
+    env: prod
+  mappings:
+    - object_match:
+        - database: sales
+          object_kind: table
+          protocol: postgres
+      add_labels:
+        env: prod
+        product: sales
+  priority: 10
+version: v1
+```
+
+Example database object, with applied labels:
+
+```yaml
+kind: db_object
```

Comment:
Instead of having a generic db_object and specifying object_kind in the spec, have you considered having separate resources like db_schema, db_table, db_row, etc.? Wonder if it would result in a cleaner resource structure, esp. because presumably different object types have different properties, etc.